### PR TITLE
JIT: Unspill normalize-on-load variables using exact small type

### LINKED ISSUE #90219: JIT: Bad codegen on win-x86
```csharp
// Generated by Fuzzlyn v1.5 on 2023-08-05 16:08:41
// Run on X86 Windows
// Seed: 7610693270284065118
// Reduced from 12.0 KiB to 2.2 KiB in 00:01:06
// Debug: Outputs 1
// Release: Outputs 2
public interface I0
{
}

public class C0 : I0
{
    public ulong F0;
    public uint F4;
    public C0(ulong f0, uint f4)
    {
        F0 = f0;
        F4 = f4;
    }
}

public class Program
{
    public static IRuntime s_rt;
    public static I0 s_1;
    public static uint s_2;
    public static byte[] s_4 = new byte[]{0};
    public static void Main()
    {
        CollectibleALC alc = new CollectibleALC();
        System.Reflection.Assembly asm = alc.LoadFromAssemblyPath(System.Reflection.Assembly.GetExecutingAssembly().Location);
        System.Reflection.MethodInfo mi = asm.GetType(typeof(Program).FullName).GetMethod(nameof(MainInner));
        System.Type runtimeTy = asm.GetType(typeof(Runtime).FullName);
        mi.Invoke(null, new object[]{System.Activator.CreateInstance(runtimeTy)});
    }

    public static void MainInner(IRuntime rt)
    {
        s_rt = rt;
        var vr8 = new C0(0, 0);
        var vr12 = s_4[0];
        vr8.F0 *= M2(vr12);
        long[] vr9 = new long[]{0};
        bool vr10 = (int)M2(0) <= vr9[0];
    }

    public static ulong M2(byte arg0)
    {
        for (int var0 = 0; var0 < 2; var0++)
        {
            var vr1 = new C0(0, 0);
            var vr3 = new C0(0, 0);
            uint vr14 = s_2;
            s_rt.WriteLine("c_0", vr14);
            uint vr15 = vr3.F4;
            s_2 = Program.M3(vr1, vr15);
            s_1 = s_1;
            s_1 = new C0(0, 0);
            s_rt.WriteLine("c_6", var0);
        }

        var vr5 = new C0(0, 1);
        uint vr18 = vr5.F4;
        arg0 = (byte)vr18;
        var vr7 = new C0(0, 1838341904U);
        if ((arg0 >= (byte)M3(vr7, 0)))
        {
            arg0 <<= 1;
        }

        s_rt.WriteLine("c_7", arg0);
        return 0;
    }

    public static uint M3(C0 argThis, uint arg0)
    {
        s_rt.WriteLine("c_0", arg0);
        return argThis.F4;
    }
}

public interface IRuntime
{
    void WriteLine<T>(string site, T value);
}

public class Runtime : IRuntime
{
    public void WriteLine<T>(string site, T value) => System.Console.WriteLine(value);
}

public class CollectibleALC : System.Runtime.Loader.AssemblyLoadContext
{
    public CollectibleALC(): base(true)
    {
    }
}
```

### PR DESCRIPTION
**Pull request by jakobbotsch**
The existing logic would sometimes unspill using the type of the local
that is being unspilled. This type is often wider than the exact small
type in the `LclVarDsc`, since NOL locals are normally expanded as
`CAST(<small type>, LCL_VAR<int>)`.

This causes problems since morph will in some cases avoid inserting
normalization for NOL locals when it has a subrange assertion available.
This optimization relies on the backend always ensuring that the local
will be normalized as part of unspilling and args homing.

Size-wise regressions are expected on xarch since the encoding of the
normalizing load is larger. However, as we have seen before, using wide
loads can cause significant store-forwarding stalls which can have large
negative impact on performance, so overall there is an expected perf
benefit of using the small loads (in addition to the correctness fix).

Fix #90219

### CODE REVIEW

**Review by sultrifork — APPROVED** [review]
LGTM

**Comment by jakobbotsch** [comment]
jitstress failures are #90144. I've opened #90376 for the libraries-jitstress failures, but there were some that look like infrastructure, so I will rerun it. I've opened #90377 for the Fuzzlyn failure.

File: src/coreclr/jit/codegenlinear.cpp [hunk 1]
```diff
@@ -1211,39 +1211,36 @@ void CodeGen::genUnspillRegIfNeeded(GenTree* tree)
...
+            // we further want to make sure we always leave a normalized value
+            // in the register.
+
+            var_types unspillType = varDsc->lvNormalizeOnLoad() ? varDsc->TypeGet() : varDsc->GetStackSlotHomeType();
+            assert(unspillType != TYP_UNDEF);
 
 #if defined(TARGET_LOONGARCH64)
             if (varTypeIsFloating(spillType) && emitter::isGeneralRegister(tree->GetRegNum()))
             {
                 spillType = spillType == TYP_FLOAT ? TYP_INT : TYP_LONG;
```
Comment (jakobbotsch) [inline src/coreclr/jit/codegenlinear.cpp:1237]:
```suggestion
                unspillType = unspillType == TYP_FLOAT ? TYP_INT : TYP_LONG;
```

**Comment by jakobbotsch** [comment]
cc @dotnet/jit-contrib PTAL @BruceForstall 

[Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=370979&view=ms.vss-build-web.run-extensions-tab). Not quite as bad as the ones I displayed during the meeting after fixing my bug, but still somewhat bad. But as mentioned this is not expected to be a perf regression, but rather an improvement (but let's see in the perf lab).

We could probably fix the size regression by not rex prefixing `movzx` unnecessarily, but I don't think it's something we need to do for .NET 8. I will open a separate issue for it.

**Comment by BruceForstall** [comment]
I was surprised to see a few examples in arm64 diffs of removing a register from byref GC reporting, but no codegen diff. The cases seemed benign since we don't need to report register pointers to the stack, but is it expected that this PR changes GC reporting?

Example:
```
@@ -436,14 +436,12 @@ G_M8490_IG31:        ; bbWeight=0, gcrefRegs=D800000 {x23 x24 x26 x27}, byrefReg
             blr     x2
             ; gcr arg pop 0
             ldr     x2, [fp, #0x18]	// [V22 tmp14]
-            ; byrRegs +[x2]
             ldr     w0, [x2, #0x04]
             cmp     w0, #24
             mov     x3, x2
             bls     G_M8490_IG05
 						;; size=68 bbWeight=0 PerfScore 0.00
 G_M8490_IG32:        ; bbWeight=0, gcrefRegs=D800000 {x23 x24 x26 x27}, byrefRegs=100000 {x20}, byref
-            ; byrRegs -[x2]
             mov     x0, x27
             ; gcrRegs +[x0]
             mov     w1, wzr
```

**Comment by jakobbotsch** [comment]
> The cases seemed benign since we don't need to report register pointers to the stack, but is it expected that this PR changes GC reporting?

Hmm no -- it happens because `GetStackSlotHomeType` uses the type from the `LclVarDsc`, while we used to use the type from the `GenTreeLclVarCommon` that we are unspilling because of. I don't think it will be a problem, but I don't think we need to take the risk for .NET 8, so let me special case this for byrefs/GC pointers to use the original behavior.

**Comment by BruceForstall** [comment]
On win-x64, the (non-GC) diffs seem to be converting from `mov DWORD` to `movzx BYTE or WORD`. But on linux-arm, it seems like the opposite:

```
-            ldrb    r10, [sp+0x14]	// [V06 tmp3]
+            ldr     r10, [sp+0x14]	// [V06 tmp3]
```

Would that be expected?

File: src/coreclr/jit/codegenlinear.cpp [hunk 2]
```diff
@@ -1211,39 +1211,36 @@ void CodeGen::genUnspillRegIfNeeded(GenTree* tree)
...
-                spillType = lclLoadType;
-            }
+            GenTreeLclVar* lcl    = unspillTree->AsLclVar();
+            LclVarDsc*     varDsc = compiler->lvaGetDesc(lcl);
+
+            // Pick type to reload register from stack with. Note that in
+            // general, the type of 'lcl' does not have any relation to the
+            // type of 'varDsc':
+            //
+            // * For NOL locals it is wider under normal circumstances, where
```
Comment (BruceForstall) [inline src/coreclr/jit/codegenlinear.cpp:1221]: nit: spell out NOL the first time
```suggestion
            // * For normalize-on-load (NOL) locals it is wider under normal circumstances, where
```

File: src/coreclr/jit/codegenlinear.cpp [hunk 3]
```diff
@@ -1211,39 +1211,36 @@ void CodeGen::genUnspillRegIfNeeded(GenTree* tree)
...
+            // morph has added a cast on top
+            // * For all locals it can be narrower narrower in some cases, e.g.
```
Comment (BruceForstall) [inline src/coreclr/jit/codegenlinear.cpp:1223]:
```suggestion
            // * For all locals it can be narrower in some cases, e.g.
```

**Comment by jakobbotsch** [comment]
> [Edit] looks like you enabled the new unspill code for `TARGET_ARM` where that target previously had different behavior.

I unified the logic because I do not think ARM32 needs or should be special cased here. But let me do that change for .NET 9 instead, and scope this change down to only normalize-on-load locals for the other architectures.

**Review by BruceForstall — APPROVED** [review]
LGTM

**Pull request merged**